packages/smithy-core/tests/unit/test_retries.py

```diff
@@ -275,3 +280,111 @@ async def test_retry_strategy_resolver_rejects_invalid_type() -> None:
         match="retry_strategy must be RetryStrategy, RetryStrategyOptions, or None",
     ):
         await resolver.resolve_retry_strategy(retry_strategy="invalid")  # type: ignore
+
+
+class TestTokenBucket:
+    @pytest.mark.asyncio
+    async def test_initial_state(self):
+        token_bucket = TokenBucket()
+        assert token_bucket.current_capacity == token_bucket.MIN_CAPACITY
+        assert token_bucket.max_capacity == token_bucket.MIN_CAPACITY
+        assert token_bucket.fill_rate == token_bucket.MIN_FILL_RATE
+
+    @pytest.mark.asyncio
+    async def test_acquire_succeeds_immediately_within_capacity(self):
+        token_bucket = TokenBucket()
+        start_time = time.monotonic()
+        await token_bucket.acquire(1)
+        elapsed = time.monotonic() - start_time
+
+        assert elapsed < 0.001  # Should be near instant
+        assert token_bucket.current_capacity == 0
+
+    @pytest.mark.asyncio
+    async def test_acquire_waits_when_capacity_insufficient(self):
+        token_bucket = TokenBucket(curr_capacity=0)
+
+        with patch("asyncio.sleep", new_callable=AsyncMock) as mock_sleep:
+
+            async def side_effect(delay: float):
+                async with token_bucket._lock:  # type: ignore
+                    token_bucket._curr_capacity = 1.0  # type: ignore
+
+            mock_sleep.side_effect = side_effect
+            await token_bucket.acquire(1)
+            assert mock_sleep.call_count == 1
+
+            actual_delay = mock_sleep.call_args[0][0]
+            assert actual_delay == pytest.approx(2.0, abs=0.05)  # type: ignore
+
+    @pytest.mark.asyncio
+    async def test_multiple_refills_over_time(self):
+        token_bucket = TokenBucket(curr_capacity=0, max_capacity=10, fill_rate=2.0)
+
+        time_values = iter([1.0, 1.5, 4.0])
+        with patch("time.monotonic", side_effect=lambda: next(time_values)):
+            token_bucket._last_timestamp = 0.0  # type: ignore
```

Comment:
We should try to use actual values for `time.monotonic()` we can "freeze" this to some fixed point in time and scale off that. Another issue here is we probably shouldn't be mutating/mocking the private vars. If the first call needs to set this to our baseline, can we just add that to the start of the list knowing it'll be a sleep "no-op"? e.g.

```suggestion
        time_values = iter([0.0, 1.0, 1.5, 4.0])
        with patch("time.monotonic", side_effect=lambda: next(time_values)):
```